### PR TITLE
[GlobalISel][AArch64] Generate ptrtoint/inttoptr as opposed to bitcast in unmerge combine.

### PR DESCRIPTION
**Pull request by davemgreen**
When combining unmerge we could end up with ptr to i64 bitcasts. Make sure they are created as ptrtoint/inttoptr instead.

### CODE REVIEW

**Comment by aemerson** [comment]
> Some of the test are still disabled as they hit other issues.

You mean existing disabled tests, hence they're unchanged by this commit?

**Comment by aemerson** [comment]
Oh I see the commented out test now.

File: llvm/test/CodeGen/AArch64/getelementptr.ll
Comment (aemerson) [inline llvm/test/CodeGen/AArch64/getelementptr.ll:1]: Do we just not have coverage for this elsewhere in the AArch64 tests?

File: llvm/test/CodeGen/AArch64/getelementptr.ll [hunk 1]
```diff
@@ -0,0 +1,454 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=aarch64 -verify-machineinstrs %s -o - | FileCheck %s --check-prefixes=CHECK,CHECK-SD
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/getelementptr.ll:2]: Use explicit -global-isel=0 for dag tests

File: llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h [hunk 2]
```diff
@@ -1218,8 +1218,14 @@ class LegalizationArtifactCombiner {
...
 
-      if (!ConvertOp && DestTy != MergeSrcTy)
-        ConvertOp = TargetOpcode::G_BITCAST;
+      if (!ConvertOp && DestTy != MergeSrcTy) {
+        if (DestTy.isPointer())
+          ConvertOp = TargetOpcode::G_INTTOPTR;
+        else if (MergeSrcTy.isPointer())
+          ConvertOp = TargetOpcode::G_PTRTOINT;
+        else
+          ConvertOp = TargetOpcode::G_BITCAST;
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h:1227]: I swear we have this exact code somewhere already

File: llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h [hunk 3]
```diff
@@ -1218,8 +1218,14 @@ class LegalizationArtifactCombiner {
     } else {
       LLT MergeSrcTy = MRI.getType(MergeI->getOperand(1).getReg());
 
-      if (!ConvertOp && DestTy != MergeSrcTy)
-        ConvertOp = TargetOpcode::G_BITCAST;
+      if (!ConvertOp && DestTy != MergeSrcTy) {
+        if (DestTy.isPointer())
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h:1222]: Mir test for this?

File: llvm/test/CodeGen/AArch64/getelementptr.ll [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+; CHECK-GI-NEXT:    sshll v1.2d, v1.2s, #0
+; CHECK-GI-NEXT:    fmov x8, d1
+; CHECK-GI-NEXT:    mov x10, v1.d[1]
+; CHECK-GI-NEXT:    mul x8, x8, x9
+; CHECK-GI-NEXT:    mul x9, x10, x11
+; CHECK-GI-NEXT:    mov v1.d[0], x8
+; CHECK-GI-NEXT:    mov v1.d[1], x9
+; CHECK-GI-NEXT:    add v0.2d, v0.2d, v1.2d
+; CHECK-GI-NEXT:    ret
+  %d = getelementptr i32, <2 x ptr> %p, i32 %q
```
Comment (madhur13490) [inline llvm/test/CodeGen/AArch64/getelementptr.ll:38]: The generated code is pretty suboptimal compared to SDAG. Just curious, do you know why?

File: llvm/test/CodeGen/AArch64/getelementptr.ll [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+; CHECK-GI-NEXT:    mov d1, v0.d[1]
+; CHECK-GI-NEXT:    ret
+  %d = getelementptr i32, ptr %p, <3 x i32> <i32 -10, i32 -10, i32 -10>
+  ret <3 x ptr> %d
+}
+
+;define <4 x ptr> @v4b_m10(ptr %p, <4 x i32> %q) {
+;  %d = getelementptr i32, ptr %p, <4 x i32> <i32 -10, i32 -10, i32 -10, i32 -10>
+;  ret <4 x ptr> %d
+;}
```
Comment (madhur13490) [inline llvm/test/CodeGen/AArch64/getelementptr.ll:454]: Remove this code?

File: llvm/test/CodeGen/AArch64/getelementptr.ll
Comment (davemgreen) [reply llvm/test/CodeGen/AArch64/getelementptr.ll:1]: Oh It turns out I've been here before and already added a ptradd.ll test. I'll switch to that one.

File: llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h (hunk 2 above)
Comment (davemgreen) [reply llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h:1227]: Yeah - there is some code in MachineIRBuilder::buildCast that does something quite similar

**Review by arsenm — APPROVED** [review]

**Pull request merged**